pinot-core/src/main/java/org/apache/pinot/core/minion/SegmentConverter.java

```diff
@@ -89,6 +95,21 @@ public SegmentConverter(List<File> inputIndexDirs, File workingDir, String table
     _recordAggregator = recordAggregator;
     _groupByColumns = groupByColumns;
     _skipTimeValueCheck = skipTimeValueCheck;
+
+    // Validate that segment schemas from all segments are the same
+    Set<Schema> schemas = new HashSet<>();
+    for (File indexDir : inputIndexDirs) {
+      try {
+        schemas.add(new SegmentMetadataImpl(indexDir).getSchema());
+      } catch (IOException e) {
+        throw new RuntimeException("Caught exception while reading schema from: " + indexDir);
+      }
+    }
+    if (schemas.size() == 1) {
+      _schema = schemas.iterator().next();
+    } else {
+      throw new IllegalStateException("Schemas from input segments are not the same");
+    }
```

Comment:
How about bailing out early from the for loop if a new schema is encountered instead of continuing to process all the files.

Reply:
Good suggestion. Submit a PR to fix it: https://github.com/apache/incubator-pinot/pull/6931